File: string/3.cpp
/*
 * @Author: httermin
 * @Date: 2019-09-09 23:37:52
 */
#include<string>
#include<iostream>
#include <unordered_map>

using namespace std;

class Solution3 {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> mymap;

        int maxLength = 0;
        int startIdx = -1;

        for(int i=0; i<s.size();i++){
            
            if(mymap.find(s[i])!=mymap.end()){
                startIdx = max(mymap[s[i]], startIdx);
                
            }
            cout<<startIdx<<endl;
            mymap[s[i]] = i;
            maxLength = max(maxLength, i - startIdx);
        }
        return maxLength;
    }
};

class SolutionOLD {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> mymap;
        unordered_map<char, int>::iterator it;
        
        int max_len = 0;
        int i = -1; 
        for(int j=0; j<s.size(); j++){
            
            it = mymap.find(s[j]);
            if(it != mymap.end()){ // 重复
                i = max(it->second, i);
            }
            
            mymap[s[j]] = j;
            max_len = max(max_len, j-i);
            
        }
        return max_len; 
    }
};

int main(){
    string s = "pwwkew";
    Solution3 ss;
    cout<<ss.lengthOfLongestSubstring(s)<<endl;
    return -1;
}

File: string/3.py
# coding=UTF-8
'''
@Author: httermin
@Date: 2019-12-23 12:39:13
'''
import sys
import os

sys.path.append("..")
class Solution(object):

    def lengthOfLongestSubstring(self, s):
        max_length = 0
        posi_dict = {}

        last_index = -1

        for i, c in enumerate(s):

            posi_c = posi_dict.get(c)

            if posi_c is not None:
                last_index = max(last_index, posi_c)
            
            max_length = max(max_length, i - last_index)
            posi_dict[c] = i 

        return max_length

    def lengthOfLongestSubstring_loop(self, s):
        """
        :type s: str
        :rtype: int
        """
        maxlength = 0
        substr = ""

        for c in s:

            try:
                position = substr.index(c)
            except ValueError:
                position = None

            if position is None:
                substr += c 
            else:
                substr = substr[position+1:] + c

            maxlength = max(maxlength, len(substr))
        return maxlength

if __name__ == "__main__":
    s = ["abcabcbb", "bbbbb", "pwwkew", "", "aabaab!bb"]

    solution = Solution()

    for t in s:
        print(solution.lengthOfLongestSubstring(t))